### ISSUE TITLE
As a software developer of scientific software, I need a native Haskell implementation of the dswap function so that I can swap elements between two vectors in a type-safe, thread safe manner.

### ISSUE THREAD

**Issue by dlewissandy** [dlewissandy/lambda-blas#69]
Documentation for the dswap function can be found at BLAS.

- [x] create native implementation of function
- [x] create haddock documentation of function
- [x] create unit test to demonstrate byte equivalence to FORTRAN BLAS library function
- [x] create benchmark(s) to quantify time performance relative to FORTRAN BLAS library function
- [x] demonstrate that native has a relative time performance of no more than 100x the FORTRAN blas function.

**Comment by dlewissandy** [comment]
Implementation Strategy:
* This function can be implemented easily by polymorphising the single precision version.

**Comment by dlewissandy** [comment]
Attached is the raw benchmark data:
[dswap.zip](https://github.com/dlewissandy/lambda-blas/files/868038/dswap.zip)

**Comment by dlewissandy** [comment]
# DSWAP BENCHMARK ANALYSIS #

## Special Case inc=1 ##
When the increment of traversal is equal to unity, the time performance of both the unsafe and pure implementations are well described by a linear model:

### PURE ###
```
lm(formula = PURE ~ N, data = dat[dat$INC == 1, ])

Residuals:
     Min       1Q   Median       3Q      Max
-11.9159  -0.6936   1.1982   1.2700  15.6580

Coefficients:
              Estimate Std. Error t value Pr(>|t|)
(Intercept) -1.1362681  0.6927693   -1.64    0.109
N            0.0085626  0.0001053   81.31   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 3.909 on 37 degrees of freedom
Multiple R-squared:  0.9944,	Adjusted R-squared:  0.9943
F-statistic:  6612 on 1 and 37 DF,  p-value: < 2.2e-16
```
### UNSAFE ###
```
lm(formula = UNSAFE ~ N, data = dat[dat$INC == 1, ])

Residuals:
    Min      1Q  Median      3Q     Max
-0.9274 -0.1713  0.1733  0.2122  1.0485

Coefficients:
              Estimate Std. Error t value Pr(>|t|)
(Intercept) -1.271e-01  6.699e-02  -1.897   0.0656 .
N            2.095e-03  1.018e-05 205.719   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.378 on 37 degrees of freedom
Multiple R-squared:  0.9991,	Adjusted R-squared:  0.9991
F-statistic: 4.232e+04 on 1 and 37 DF,  p-value: < 2.2e-16
```
### ASYMPTOTIC PERFORMANCE ###
In the limit as n becomes very large, the relative time performance should approach:
```
0.0085626
-------- = 4.1x
2.095e-3
```
which is within the 100x performance threshold for this project milestone

## Special Case inc/=1 ##
When the increment of traversal is not equal to unity, the time performance of both the unsafe and pure implementations are well described by a multi-linear model:

### PURE ###
```
lm(formula = PURE ~ N + Z, data = dat[dat$INC != 1, ])

Residuals:
    Min      1Q  Median      3Q     Max
-1.5123 -0.5542  0.2553  0.5610  1.6729

Coefficients:
              Estimate Std. Error t value Pr(>|t|)
(Intercept) -4.278e-01  1.206e-01  -3.546 0.000884 ***
N            1.848e-02  2.814e-04  65.659  < 2e-16 ***
Z            1.597e-03  2.151e-05  74.235  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.7332 on 48 degrees of freedom
Multiple R-squared:  0.9986,	Adjusted R-squared:  0.9986
F-statistic: 1.762e+04 on 2 and 48 DF,  p-value: < 2.2e-16
```
### UNSAFE ###
```
lm(formula = UNSAFE ~ N + Z, data = dat[dat$INC != 1, ])

Residuals:
    Min      1Q  Median      3Q     Max
-1.2687 -0.3270  0.1990  0.3746  1.0855

Coefficients:
              Estimate Std. Error t value Pr(>|t|)
(Intercept) -3.226e-01  8.506e-02  -3.792 0.000418 ***
N            7.609e-03  1.984e-04  38.343  < 2e-16 ***
Z            9.914e-04  1.517e-05  65.369  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.517 on 48 degrees of freedom
Multiple R-squared:  0.9976,	Adjusted R-squared:  0.9975
F-statistic:  9782 on 2 and 48 DF,  p-value: < 2.2e-16
```
### ASYMPTOTIC PERFORMANCE ###
In the limit as both n and inc become very large, the relative time performance should approach:
```
1.597e-03
--------- = 1.6x
9.914e-04
```
which is within the 100x performance threshold for this project milestone.

**Comment by dlewissandy** [comment]
Closed as DONE after approval of pull request #80.
